src/app/controllers/sap.php

```diff
@@ -68,6 +68,53 @@ public function logout() {
         $this->sap_auth->logout();
     }
 
+    public function verify($hash="") {
+        $this->load_library('http_lib');
+        if ($hash == "") {
+            $this->http_lib->response_code(404);
+        }
+        $this->load_model('sap_model');
+        $user = $this->sap_model->verify_hash($hash);
+        if (! $user) {
+            $this->http_lib->response_code(404);
+        }
+        $this->load_library('csrf_lib');
+        $this->load_library('session_lib');
+        $this->load_view('sap/create_password', [
+            'csrf_token' => $this->csrf_lib->new_csrf_token(),
+            'hash' => $hash,
+            'error' => $this->session_lib->flash_get('error'),
```

Comment:
Yeah, this is the link that goes in the email. The error gets set in line 113, when the user tries to set a password, but the application fails to set it.